Add support for raw (non-formatted) volume

slaperche-scality opened the pull request:
**Component**:

operator, salt

**Context**: 

We want to support the provisioning of non-formatted block device.

**Summary**:

Add a new parameter `noFormat` in the Volume CRD and update Salt code to adapt the provisioning accordingly.

**Acceptance criteria**: 

Tests are passing (`sparseLoopDevice`)
Manual testing (`rawBlockDevice`)

---

Closes: #2421, #2963

slaperche-scality committed:
doc/design: move volume with other architecture docs

Having a set of "KEP-style" docs is too heavy for our needs.

Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/types: add noFormat option

This options allows to skip the formatting of the device.
It's optional and will be set to false by default (i.e., format the
device) to preserve the backward compatibility.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/controller: set PV mode according to the new option

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: add a new module function to retrieve size & path

This will replace call to `disk.dump`.
By returning both size and path, we can keep a single Salt call for
unformatted volume (for which we will need to get a reliable path from
Salt).

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/controller: use device_info instead of disk.dump

Since we will now need more information than just the size, we replace
the call to `disk.dump` by a call to our function
`metalk8s_volumes.device_info`.

For now, we simply plug the new call, we don't use information other
than size yet.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt: add pillar refresh to volumes.prepared

We now call `metalk8s_volumes.device_info` instead of `disk.dump`.
Unlike the later, this new function relies on the pillar content and we
have a problem here because calling a Salt module doesn't refresh the
pillar.

That being said, the state we call before (volumes.prepared) does
automatially refresh the pillar because it works fine and it does look
up the volume from the pillar.
But it seems this pillar isn't kept around, since calling the module
right after fails.
If we explicitely refresh the pillar during our state, then it's kept
around and our module can reuse the fetched pillar: that's why we add a
refresh_pillar call to our state.

Now, in the unlikely case that we have a crash/restart after the state
execution BUT before executing the module, we won't have access to a
cached pillar: in such case we have a last-chance try/except that
refresh the pillar from inside the module.

I'm not super happy with the current approach…, but I don't see a better
one:
- using a state is not possible (because I want to return data)
- calling `saltutil.refresh_pillar` from the operator will complicate
  the code a bit and won't even solve the "I have a crash/reboot/restart
  just before calling my module and thus no cached pillar!" situation.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/controller: use a volatile cache for device info

Instead of persisting device information into the status (which is
hackish), we now store them in a hashmap.
It will be good enough for most of the cases, and in the worst case (we
crash after getting the device info but before creating the PV), we can
simply reschedule the Salt call to fetch the device information.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: revert the Result field

We no longer use it (since we're using a volatile cache in the
operator).
We still keep some compatibility with this format, in case we have
volume with serialized Result.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: add support for non-formatted sparseLoopDevice

We create a new class, SparseLoopDeviceNoFormat, that inherit from
SparseLoopDevice and customize the formatting step.

BTW, we should probably rename this step, as it's no longer about
formatting in every cases.

Note that because the Salt module `parted_partition` doesn't work on
sparse file, we call `parted` directly.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: s/format/prepare/

The `format` step is no longer only about formatting (for noFormat)
volume, so let's rename it to `prepare` to avoid confusion.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: add support for non-formatted rawBlockDevice

We have to handle different cases here, the device path can points to:
- a LVM volume: we have nothing to do (we'll use LVM UUID)
- a disk partition (we make sure the partition's label is the volume UUID)
- an entire disk: we create a GPT table and a single partition with the
  right label.

Since the last case is the same as SparseLoopDeviceNoFormat, we factor
out the common code into `prepare_noformat`.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: update device_info to handle noformat volume

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: add test for noformat volume

Closes: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/feature/2421-add-support-for-raw-block-volume` with contents from `feature/2421-add-support-for-raw-block-volume`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/feature/2421-add-support-for-raw-block-volume origin/development/2.6
 $ git merge origin/feature/2421-add-support-for-raw-block-volume
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/feature/2421-add-support-for-raw-block-volume
```

slaperche-scality commented:
Will do some more manual testing, but code is ready for review

slaperche-scality committed:
doc/operation: document the `noFormat` parameter

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/doc: add comment about `deviceName` refresh

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
doc/design: update volume design

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

gdemonet commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -48,4 +48,11 @@ Provision backing storage for {{ volume }}:
 
+Update pillar after volume provisionning:
+  module.run:
+    - saltutil.refresh_pillar:
+      - wait: True
```
To make this call (and the above) synchronous, you'll need to explicitly pass `async=False`.
For reference, see:
https://github.com/saltstack/salt/issues/20590
https://github.com/saltstack/salt/pull/50168

gdemonet replied:
Which makes me think there are some workarounds we could remove now that we upgraded to Salt 3000 (cc @TeddyAndrieux)

slaperche-scality replied:
Ho good catch, will add `async=False` then.

slaperche-scality replied:
Will revert this for now, seems not supported (or I got the syntax wrong):

```
      message: 'Salt job ''PrepareVolume'' failed with: ''saltutil.refresh_pillar''
        failed: refresh_pillar() got an unexpected keyword argument ''async'''
```

gdemonet replied:
No indeed my bad, it seems the `wait: True` is the most recent implementation.

gdemonet commented on docs/developer/architecture/volume.rst:
```diff
@@ -148,6 +148,28 @@ Similarly, our **Volume** object will have the following states:
+  * the **rawBlockDevice** is a partition (e.g. ``/dev/sdb1``): we re-label the
+    partition using the **Volume** UUID and use
+    ``/dev/disk/by-partlabel/<volume-uuid>`` as device path.
+  * The **rawBlockDevice** is a LVM volume: we use the exitisting LVM UUID and
```
```suggestion
  * The **rawBlockDevice** is a LVM volume: we use the existing LVM UUID and
```

slaperche-scality replied:
Done

gdemonet commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -451,6 +451,15 @@ def __init__(self, volume):
+        if self._kind == DeviceType.LVM:
+            for symlink in glob.glob('/dev/disk/by-id/dm-uuid-LVM-*'):
+                realpath = os.path.realpath(symlink)
+                if os.path.basename(realpath) == name:
```
IIUC, `name` should be computed (maybe make this an attribute or a dynamic property?):

```suggestion
                if os.path.basename(realpath) == device_name(self.path):
```

slaperche-scality replied:
Indeed, `name` is not defined >_<

Fixed.

gdemonet reviewed (commented):
Some early comments, most of it looks good to me :)

slaperche-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/feature/2421-add-support-for-raw-block-volume` with contents from `feature/2421-add-support-for-raw-block-volume`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/feature/2421-add-support-for-raw-block-volume origin/development/2.6
 $ git merge origin/feature/2421-add-support-for-raw-block-volume
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/feature/2421-add-support-for-raw-block-volume
```

slaperche-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/feature/2421-add-support-for-raw-block-volume` with contents from `feature/2421-add-support-for-raw-block-volume`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/feature/2421-add-support-for-raw-block-volume origin/development/2.6
 $ git merge origin/feature/2421-add-support-for-raw-block-volume
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/feature/2421-add-support-for-raw-block-volume
```

slaperche-scality committed:
salt/volume: add retry logic to device name resolution

It seems that, in some cases, you may have a small delay where the
partition is not visible under `/dev` after some operation (like
changing the label):

    [root@test-master-1 centos]# ls /dev | grep vdb
    vdb
    vdb1
    [root@test-master-1 centos]# parted /dev/vdb name 1 foo && ls /dev/ | grep vdb
    vdb
    [root@test-master-1 centos]# ls /dev | grep vdb
    vdb
    vdb1

Because of that, device name resolution may suffer transient failure.
So let's retry a few times before giving up.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on docs/operation/volume_management/volume_creation_deletion_cli.rst:
```diff
@@ -26,13 +26,16 @@ This section describes how to create a **Volume** from the **CLI**.
    - **nodeName**: the name of the node where the volume will be located.
    - **storageClassName**: the **StorageClass** to use
    - **devicePath**: path to the block device (for example, `/dev/sda1`).
+   - **noFormat**: skip the volume formatting if **true**, optional (default to
```
Should `noFormat` be inside the `rawBlockDevice` section? Why not top-level?

Also, instead of `noFormat`, could we re-use the `volumeMode` as used `PersistentVolume(Claim)`?

slaperche-scality replied:
Yeah, good point.

I'll add a commit to change that.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -180,6 +182,32 @@ def device_name(path):
+    """Return information about the backing storage device of the given volume.
+
+    Returned information are:
+    - an reliable path (e.g.: constant accross reboot) to the storage device
```
```suggestion
    - a stable path (e.g.: constant across reboots) to the storage device
```

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -339,6 +378,27 @@ def clean_up(self):
+        """Check if the volume is already prepared."""
+        # Partition 1 should be labelled with the volume UUID.
+        pattern = '1.+{}'.format(self.get('metadata.uid').lower())
+        result  = _run_cmd(' '.join(['parted', self.path, 'print']))
```
This requires `parted` to be installed. Can we build this on top of plain `sfdisk` instead?

slaperche-scality replied:
From `man 8 sfdisk`:

> sfdisk doesn't understand the GUID Partition Table (GPT) format and it is not designed for large partitions.  In these cases use the more advanced GNU parted(8).

So I guess `parted` is a safer choice (will need to install it through Salt to be sure it's present though).

NicolasT replied:
Modern `fdisk` does know GPT, but maybe not on CentOS 7. Anyway, on those systems there should be `sgdisk` for `fdisk`-like GPT management.

slaperche-scality replied:
```
[root@test-bootstrap centos]# sgdisk
bash: sgdisk: command not found
```

Seems not installed by default, so what's the advantage of using it instead of `parted` (since I'll have to make sure it's installed through Salt anyway)?

NicolasT replied:
Fair enough. I was thinking `parted` has (many) more deps than `gdisk`, but that seems not to be the case (they're comparable), so if `parted` does what you need, go for it.

slaperche-scality replied:
Done, I'm now using `sgdisk` since it's possible to set the partition GUID with it.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -377,6 +437,79 @@ def is_cleaned_up(self):
+        if self._kind == DeviceType.LVM:
+            return
+        uuid = self.get('metadata.uid')
+        # In case of partition, set the label to volume UUID.
```
Setting the *label* to the volume UUID? Can't we change the partition UUID instead?

slaperche-scality replied:
Yeah, I wanted to use partition UUID at first, but couldn't find a way to set it :confused:  so I went with partition label instead.

NicolasT replied:
Not fond of that at all, because super confusing. Instead, I'd set the *name* of the GPT partition to the `metadata.name` of the `Volume` object...

slaperche-scality replied:
I used UUID for consistency with the others, but I can change it to `name`.

NicolasT replied:
I'm not sure we're aligned here :)

In GPT, a partition can have a *name* and a *GUID*. So, what I suggest is to set its *GUID* to the `metadata.uid` of the `Volume` object, and the *name* to `metadata.name` of the `Volume`. Then, always use the `/dev/disk/by-partuuid/...` path to *refer* to it. The *name* is merely cosmetic.

slaperche-scality replied:
Yup, but:

> I wanted to use partition UUID at first, but couldn't find a way to set it :confused: 

Haven't find a way to set the partition UID with parted.
Will give a try to `sgdisk`, maybe it has more options.

slaperche-scality replied:
Easy to do with `sgdisk`, I've added a new commit to use `sgdisk` instead of `parted`.

NicolasT reviewed (dismissed):
Some questions and suggestions.

I think we should target `sfdisk` instead of `parted` to 'get things done', if feasible...

slaperche-scality committed:
salt/volume: make sure parted is installed

`parted` is used to deal with GPT partitions.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator: do not always update LastUpdateTime

In some case, this trigger an infinite reconciliation loop…

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator: use corev1.PersistentVolumeMode instead of a boolean

This makes Volume closer to the underlying PersistentVolume (which
simplify the code and also helps the user already familiar with
PersistentVolume).

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: use volume name for GTP label instead of UUID

Will be less confusing.

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: use sgdisk instead of parted

By using sgdisk we can change the partition GUID to match the volume
UUID. Haven't found how to do it with parted…

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
Ready for a new round of review, I've applied all the requested changes and tested manually the different cases for Block mode (whole disk, partition and LVM volume) and they work.

That being said, unlike Filesystem mode, we now some distinct codepaths for Block mode (partition and LVM cases are specific to `rawBlockDevice`): at one point, we may want to have tests in the CI for `rawBlockDevice` as well.

slaperche-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/feature/2421-add-support-for-raw-block-volume` with contents from `feature/2421-add-support-for-raw-block-volume`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/feature/2421-add-support-for-raw-block-volume origin/development/2.6
 $ git merge origin/feature/2421-add-support-for-raw-block-volume
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/feature/2421-add-support-for-raw-block-volume
```

alexandre-allard-scality commented on docs/developer/architecture/volume.rst:
```diff
@@ -157,6 +148,28 @@ Similarly, our **Volume** object will have the following states:
+  ``dev/disk/by-uuid/<volume-uuid>`` as device path.
+* **sparseLoopDevice** without filesystem: we create a GUID Partition Table on
+  the sparse file and create a single partition encompassing the whole device,
+  setting the GUID of the partition to the **Volume** UUID. We can then use
```
Should be `the UUID of the partition`, no ? Or maybe I'm lost with all these UUID/GUID. 
```suggestion
  setting the UUID of the partition to the **Volume** UUID. We can then use
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

alexandre-allard commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -377,6 +442,76 @@ def is_cleaned_up(self):
+        super(RawBlockDeviceBlock, self).__init__(volume)
+        # Detect which kind of device we have: a real disk, only a partition or
+        # an LVM volume.
+        name = device_name(self.path)
+        self._partition = None
+        for symlink in glob.glob('/dev/disk/by-id/dm-uuid-LVM-*'):
+            realpath = os.path.realpath(symlink)
+            if os.path.basename(realpath) == name:
+                self._kind = DeviceType.LVM
+                return
```
Minor, it looks like the code to find a LVM symlink here and in `persistent_path()` could be factorized in a function:
```
def get_lvm_path(self):
    name = device_name(self.path)
    for symlink in glob.glob('/dev/disk/by-id/dm-uuid-LVM-*'):
        realpath = os.path.realpath(symlink)
        if os.path.basename(realpath) == name:
            return symlink
    return None
```
```suggestion
        self._partition = None
        if self.get_lvm_path():
            self._kind = DeviceType.LVM
            return
```

slaperche-scality replied:
Good point, I've added a commit to factorize it!

alexandre-allard reviewed (dismissed):
LGTM.

slaperche-scality committed:
salt/volume: factor out the LVM lookup logic

Refs: #2421
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
